presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -291,6 +415,98 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
         return Math.floor(num * factor + 0.5) / factor;
     }
 
+    private static SqlScalarFunction decimalRoundNFunction()
+    {
+        Signature signature = Signature.builder()
+                .kind(SCALAR)
+                .name("round")
+                .literalParameters("num_precision", "num_scale")
+                .argumentTypes("decimal(num_precision, num_scale)", StandardTypes.BIGINT)
+                .returnType("decimal(num_precision, num_scale)")
+                .build();
+        return SqlScalarFunction.builder(MathFunctions.class)
+                .signature(signature)
+                .description("remainder of given quotient")
+                .methods("roundShortDecimal", "roundLongDecimal")
+                .extraParameters(MathFunctions::decimalRoundNExtraParameters)
+                .build();
+    }
+
+    private static SqlScalarFunction decimalRoundFunction()
+    {
+        Signature signature = Signature.builder()
+                .kind(SCALAR)
+                .name("round")
+                .literalParameters("num_precision", "num_scale")
+                .argumentTypes("decimal(num_precision, num_scale)")
+                .returnType("decimal(num_precision, num_scale)")
+                .build();
+        return SqlScalarFunction.builder(MathFunctions.class)
+                .signature(signature)
+                .description("remainder of given quotient")
+                .methods("roundShortDecimal", "roundLongDecimal")
+                .extraParameters(MathFunctions::decimalRoundExtraParameters)
+                .build();
+    }
+
+    private static List<Object> decimalRoundExtraParameters(SpecializeContext specializeContext)
+    {
+        return ImmutableList.of(0, decimalRoundNExtraParameters(specializeContext).get(0), decimalRoundNExtraParameters(specializeContext).get(1));
+    }
+
+    private static List<Object> decimalRoundNExtraParameters(SpecializeContext context)
+    {
+        return ImmutableList.of(context.getLiteral("num_precision"), context.getLiteral("num_scale"));
+    }
+
+    public static long roundShortDecimal(long num, long roundScale, long inputPrecision, long inputScale)
+
+    {
+        if (num == 0 || inputPrecision - inputScale + roundScale <= 0) {
+            return 0;
+        }
+        if (roundScale >= inputScale) {
+            return num;
+        }
+        if (num < 0) {
+            return -roundShortDecimal(-num, roundScale, inputPrecision, inputScale);
+        }
+
+        long rescaleFactor = tenToNth((int) (inputScale - roundScale)).longValueExact();
```

Comment:
tenToNth is just a table lookup, but OK, since we already decided to break those two implemenations apart I'll pre compute this in specialize